File: ProjectTracker20vb/ProjectTracker.Library/ResourceAssignments.vb
<Serializable()> _
Public Class ResourceAssignments
  Inherits BusinessListBase(Of ResourceAssignments, ResourceAssignment)

#Region " Business Methods "

  Default Public Overloads ReadOnly Property Item(ByVal projectId As Guid) As ResourceAssignment
    Get
      For Each res As ResourceAssignment In Me
        If res.ProjectId.Equals(projectId) Then
          Return res
        End If
      Next
      Return Nothing
    End Get
  End Property

  Public Sub AssignTo(ByVal projectId As Guid)

    If Not Contains(projectId) Then
      Dim project As ResourceAssignment = ResourceAssignment.NewResourceAssignment(projectId)
      Me.Add(project)

    Else
      Throw New InvalidOperationException("Resource already assigned to project")
    End If

  End Sub

  Public Overloads Sub Remove(ByVal projectId As Guid)

    For Each res As ResourceAssignment In Me
      If res.ProjectId.Equals(projectId) Then
        Remove(res)
        Exit For
      End If
    Next

  End Sub

  Public Overloads Function Contains(ByVal projectId As Guid) As Boolean

    For Each project As ResourceAssignment In Me
      If project.ProjectId = projectId Then
        Return True
      End If
    Next
    Return False

  End Function

  Public Overloads Function ContainsDeleted(ByVal projectId As Guid) As Boolean

    For Each project As ResourceAssignment In DeletedList
      If project.ProjectId = projectId Then
        Return True
      End If
    Next
    Return False

  End Function

#End Region

#Region " Factory Methods "

  Friend Shared Function NewResourceAssignments() As ResourceAssignments

    Return New ResourceAssignments

  End Function

  Friend Shared Function GetResourceAssignments( _
          ByVal dr As SafeDataReader) As ResourceAssignments

    Return New ResourceAssignments(dr)

  End Function

  Private Sub New()

    MarkAsChild()

  End Sub

  Private Sub New(ByVal dr As SafeDataReader)

    MarkAsChild()
    Fetch(dr)

  End Sub

#End Region

#Region " Data Access "

  Private Sub Fetch(ByVal dr As SafeDataReader)

    Me.RaiseListChangedEvents = False
    While dr.Read()
      Me.Add(ResourceAssignment.GetResourceAssignment(dr))
    End While
    Me.RaiseListChangedEvents = True

  End Sub

  Friend Sub Update(ByVal cn As SqlConnection, ByVal resource As Resource)

    Me.RaiseListChangedEvents = False
    ' update (thus deleting) any deleted child objects
    For Each item As ResourceAssignment In DeletedList
      item.DeleteSelf(cn, resource)
    Next
    ' now that they are deleted, remove them from memory too
    DeletedList.Clear()

    ' add/update any current child objects
    For Each item As ResourceAssignment In Me
      If item.IsNew Then
        item.Insert(cn, resource)

      Else
        item.Update(cn, resource)
      End If
    Next
    Me.RaiseListChangedEvents = True

  End Sub

#End Region

End Class

File: ProjectTrackercs/ProjectTracker.Library/ResourceAssignments.cs
using System;
using System.Data;
using System.Data.SqlClient;
using Csla;
using Csla.Data;

namespace ProjectTracker.Library
{
  [Serializable()]
  public class ResourceAssignments : BusinessListBase<ResourceAssignments, ResourceAssignment>
  {
    #region Business Methods

    public ResourceAssignment this[Guid projectId]
    {
      get
      {
        foreach (ResourceAssignment res in this)
          if (res.ProjectId.Equals(projectId))
            return res;
        return null;
      }
    }

    public void AssignTo(Guid projectId)
    {
      if (!Contains(projectId))
      {
        ResourceAssignment project = ResourceAssignment.NewResourceAssignment(projectId);
        this.Add(project);
      }
      else
        throw new InvalidOperationException("Resource already assigned to project");
    }

    public void Remove(Guid projectId)
    {
      foreach (ResourceAssignment res in this)
      {
        if (res.ProjectId.Equals(projectId))
        {
          Remove(res);
          break;
        }
      }
    }

    public bool Contains(Guid projectId)
    {
      foreach (ResourceAssignment project in this)
        if (project.ProjectId == projectId)
          return true;
      return false;
    }

    public bool ContainsDeleted(Guid projectId)
    {
      foreach (ResourceAssignment project in DeletedList)
        if (project.ProjectId == projectId)
          return true;
      return false;
    }

    #endregion

    #region Factory Methods

    internal static ResourceAssignments NewResourceAssignments()
    {
      return new ResourceAssignments();
    }

    internal static ResourceAssignments GetResourceAssignments(SafeDataReader dr)
    {
      return new ResourceAssignments(dr);
    }

    private ResourceAssignments()
    {
      MarkAsChild();
    }

    private ResourceAssignments(SafeDataReader dr)
    {
      MarkAsChild();
      Fetch(dr);
    }

    #endregion

    #region Data Access

    private void Fetch(SafeDataReader dr)
    {
      RaiseListChangedEvents = false;
      while (dr.Read())
        this.Add(ResourceAssignment.GetResourceAssignment(dr));
      RaiseListChangedEvents = true;
    }

    internal void Update(Resource resource)
    {
      RaiseListChangedEvents = false;
      // update (thus deleting) any deleted child objects
      foreach (ResourceAssignment item in DeletedList)
        item.DeleteSelf(resource);
      // now that they are deleted, remove them from memory too
      DeletedList.Clear();

      // add/update any current child objects
      foreach (ResourceAssignment item in this)
      {
        if (item.IsNew)
          item.Insert(resource);
        else
          item.Update(resource);
      }
      RaiseListChangedEvents = true;
    }

    #endregion
  }
}